### DISCUSSION TITLE
Unique adjacent lines

### DISCUSSION THREAD

**Discussion by yunga** [qcz/vscode-text-power-tools#72] [Ideas]
Hello,

Along with the "Remove duplicated lines" feature, it would be useful to have basic [uniq](https://manpages.ubuntu.com/manpages/lunar/en/man1/uniq.1.html) functionalities, it's a tool to work on repeated consecutive lines. Essentials are:
- keep unique lines
- keep duplicate lines
- case sensitive / insensitive
- bonus: count the number of consecutive identical lines, padded for easy sorting

it comes handy, in conjunction with sort, to examine and manipulate text files.

---

just in case you're interested, here's a few examples working with the following `input.txt` file:

```
Ape
Ape
Ape
Bat
Cat
Cat
Dog
Cat
Cat
Cat
Eel
Fox
Fox
Fox
Fox
Fox
Fox
Fox
Fox
Fox
Fox
Fox
```

notice how `uniq input.txt` would keep the two, non-adjacent, "Cat" lines:

```
Ape
Bat
Cat
Dog
Cat
Eel
Fox
```

`uniq -d input.txt`, only shows duplicate, again "Cat" is shown twice because they're not adjacent:

```
Ape
Cat
Cat
Fox
```

finally `uniq -c input.txt` count the number of time each line appears:

```
      3 Ape
      1 Bat
      2 Cat
      1 Dog
      3 Cat
      1 Eel
     11 Fox
```

Voilà. People generally use sort and uniq at the command line, but it would be nice to be able to do it directly from vscode.

**Comment by qcz** [comment]
Added the following new commands to 1.42.0:
* _Remove duplicated lines (case insensitive)_
* _Keep only duplicated lines_
* _Keep only duplicated lines (case insensitive)_

Added the following adjacency related commands to 1.42.0:
* _Remove adjacent duplicated lines_
* _Remove adjacent duplicated lines (case insensitive)_
* _Keep only adjacent duplicated lines_
* _Keep only adjacent duplicated lines (case insensitive)_
* _Count adjacent duplicated lines_

Will implement _Keep only unique lines_ in the future. And counting commands in case insensitive mode are missing, not sure that there is a demand for it (let me know if there is).

**Reply by yunga** [reply]
Wow, thank you for implementing these features so quickly. I really appreciate your work and dedication to this extension.

`Keep only unique lines` feature will also be useful to do some simple set operations on text files.

I’m not sure how most people use count regarding case insensitivity, but I do use it. However, I don’t mind lowercasing the whole file before, it's just a keystroke away.

Another thing that I often need after a count is to sort the lines by the count index. One solution would be to pad the number with spaces or zeroes, so regular string sorting would work. Or, since this extension already has an extended choice of sort functions, maybe you could add a function to sort lines by taking the first word as a number. I don’t know what would be easier/better for you. (And feel free to ignore all this if you are too busy).

Again, thank you for your work and promptness.

**Reply by qcz** [reply]
Added in 1.43:
- Keep unique lines
- Padded counters in the count occurrences commands, so the result can be sorted.

**Reply by yunga** [reply]
There's little bug when i try to use `Keep only unique lines`, it gives this error message: `Internal error: Command called in an unsupported configuration: runKeepOnlyCommand, what: uniques, onlyAdjacent: true`. I tried with the saved example file provided above, with a new sample file or an empty file, while selecting everything, or without selecting anything.

On the other hand, the padding for count is really great that way.


Thank you for your work and the time you took to look and implement these suggestions.

**Reply by qcz** [reply]
> There's little bug when i try to use Keep only unique lines, it gives this error message: Internal error: Command called in an unsupported configuration: runKeepOnlyCommand, what: uniques, onlyAdjacent: true. I tried with the saved example file provided above, with a new sample file or an empty file, while selecting everything, or without selecting anything.

Sorry, I forgot to remove the testing for the unsupported internal combination. Fixed in 1.43.1.

> Thank you for your work and the time you took to look and implement these suggestions.

Thank you for the nice ideas.

**Reply by yunga** [reply]
It's Perfect! 👍
